In [3]:
import tensorflow as tf 
from tensorflow import keras 
from keras import Sequential
from keras.layers import Conv2D,Dense,MaxPooling2D,Flatten
from keras.activations import relu,sigmoid

In [4]:
# now create cnn models:

model = Sequential()


# convolution layer:

model.add(Conv2D(filters=32,kernel_size=(3,3),activation=relu,padding='valid',input_shape=(256,256,3)))
model.add(MaxPooling2D(pool_size=(2,2),strides=2,padding='valid'))


model.add(Conv2D(filters=64,kernel_size=(3,3),padding='valid',activation=relu))
model.add(MaxPooling2D(pool_size=(2,2),strides=2,padding='valid'))


model.add(Conv2D(filters=128,kernel_size=(3,3),padding='valid',activation=relu))
model.add(MaxPooling2D(pool_size=(2,2),strides=2,padding='valid'))

# flatten 

model.add(Flatten())

# fully connected layer:

model.add(Dense(128,activation=relu))
model.add(Dense(64,activation=relu))
model.add(Dense(1,activation=sigmoid))



/home/yasin/allProgram/100DaysDL/code/venvML/lib/python3.12/site-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
2024-05-23 12:25:29.127992: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 58982400 exceeds 10% of free system memory.
2024-05-23 12:25:29.229618: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 58982400 exceeds 10% of free system memory.
2024-05-23 12:25:29.269627: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 58982400 exceeds 10% of free system memory.


In [5]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 254, 254, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 127, 127, 32)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 125, 125, 64)   │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 62, 62, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 60, 60, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 30, 30, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 115200)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │    14,745,728 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 14,847,297 (56.64 MB)

 Trainable params: 14,847,297 (56.64 MB)

 Non-trainable params: 0 (0.00 B)

Sure! Here's the formula to calculate the number of parameters for different types of layers in a Convolutional Neural Network (CNN):

### 1. Convolutional Layers (Conv2D)
For a Conv2D layer, the number of parameters can be calculated using the following formula:

\[
 {Parameters} = (f_h * f_w * c_{in} + 1) * c_{out}
\]

where:
- \(f_h\) is the filter height
- \(f_w\) is the filter width
- \(c_{in}\) is the number of input channels (depth of the input volume)
- \(c_{out}\) is the number of output channels (number of filters)
- The \(+1\) accounts for the bias term for each filter.

### 2. Dense (Fully Connected) Layers
For a Dense layer, the number of parameters can be calculated using the following formula:

\[
\text{Parameters} = (n_{in} \times n_{out}) + n_{out}
\]

where:
- \(n_{in}\) is the number of input units
- \(n_{out}\) is the number of output units
- The \(+ n_{out}\) accounts for the bias term for each output unit.

### 3. Pooling Layers (MaxPooling2D, AveragePooling2D)
Pooling layers do not have any parameters as they simply perform a down-sampling operation.

### Example Calculation
Let's break down the calculation of parameters for each layer using these formulas for the given model:

1. **Conv2D Layer 1**
   - Filters: 32
   - Kernel size: (3,3)
   - Input shape: (256,256,3)
   - Parameters:
     \[
     \text{Parameters} = (3 \times 3 \times 3 + 1) \times 32 = (27 + 1) \times 32 = 28 \times 32 = 896
     \]

2. **Conv2D Layer 2**
   - Filters: 64
   - Kernel size: (3,3)
   - Input shape: (127,127,32)
   - Parameters:
     \[
     \text{Parameters} = (3 \times 3 \times 32 + 1) \times 64 = (288 + 1) \times 64 = 289 \times 64 = 18496
     \]

3. **Conv2D Layer 3**
   - Filters: 128
   - Kernel size: (3,3)
   - Input shape: (62,62,64)
   - Parameters:
     \[
     \text{Parameters} = (3 \times 3 \times 64 + 1) \times 128 = (576 + 1) \times 128 = 577 \times 128 = 73856
     \]

4. **Dense Layer 1**
   - Units: 128
   - Input size: 115200 (flattened from previous layer)
   - Parameters:
     \[
     \text{Parameters} = (115200 \times 128) + 128 = 14745600 + 128 = 14745728
     \]

5. **Dense Layer 2**
   - Units: 64
   - Input size: 128
   - Parameters:
     \[
     \text{Parameters} = (128 \times 64) + 64 = 8192 + 64 = 8256
     \]

6. **Dense Layer 3**
   - Units: 1
   - Input size: 64
   - Parameters:
     \[
     \text{Parameters} = (64 \times 1) + 1 = 64 + 1 = 65
     \]

By following these formulas, you can calculate the number of parameters for each layer in a CNN.